File: old-node-server/src/service/auth-service.ts
import Nedb from "@seald-io/nedb";
import * as ExpressCore from "express-serve-static-core";
import constants from "../constant/common-constants.js";
import { DatabaseEngine } from "../lib/database-engine.js";
import {
  DeveloperError,
  throwOnFalsy,
  throwOnTruthy,
  UserError
} from "../utility/coded-error.js";

export class AuthService {
  db: Nedb;

  constructor(dbEngine: DatabaseEngine) {
    this.db = dbEngine.connection;
  }

  async authenticate(_expressRequest: ExpressCore.Request) {
    const authorizationHeader = String(
      _expressRequest.headers["authorization"] || ""
    );

    throwOnFalsy(
      UserError,
      authorizationHeader.length > 0,
      "AUTHORIZATION_HEADER_MISSING",
      "Authorization header is missing"
    );

    let parts = authorizationHeader.split(" ");
    throwOnFalsy(
      UserError,
      parts.length === 2 &&
      parts[0].toLowerCase().indexOf("bearer") === 0 &&
      parts[1].length === constants.iam.API_KEY_LENGTH,
      "AUTHORIZATION_HEADER_MALFORMATTED",
      "Authorization header is malformatted"
    );
    let apiKey = parts.pop() as string;

    let session = await dispatch.sessionService.getSessionByApiKey(
      apiKey
    );

    throwOnFalsy(
      UserError,
      session,
      "API_KEY_NOT_FOUND",
      "Your session could not be found. Login again."
    );

    throwOnTruthy(
      UserError,
      session.hasExpired,
      "API_KEY_EXPIRED",
      "Your session has expired. Login again."
    );

    let hasExpired =
      Date.now() - session.createdAt >
      constants.iam.SESSION_VALIDITY_DURATION_MS;
    throwOnTruthy(
      UserError,
      hasExpired,
      "API_KEY_EXPIRED",
      "Your session has expired. Login again."
    );

    let { userId, _id: sessionId } = session;

    let user = await dispatch.userService.findUserByIdOrFail(userId);

    return { apiKey, userId, sessionId, user };
  }
}

File: web-server/internal/service/auth_service.go
package service

import (
	"context"
	"strings"
	"time"

	"github.com/nkrypt-xyz/nkrypt-xyz-web-server/internal/config"
	"github.com/nkrypt-xyz/nkrypt-xyz-web-server/internal/model"
	"github.com/nkrypt-xyz/nkrypt-xyz-web-server/internal/pkg/apperror"
)

type AuthService struct {
	sessionSvc *SessionService
	userSvc    *UserService
	cfg        *config.Config
}

func NewAuthService(sessionSvc *SessionService, userSvc *UserService, cfg *config.Config) *AuthService {
	return &AuthService{
		sessionSvc: sessionSvc,
		userSvc:    userSvc,
		cfg:        cfg,
	}
}

// Authenticate validates the Authorization header and returns auth data for the authenticated user.
func (a *AuthService) Authenticate(ctx context.Context, authorizationHeader string) (*model.AuthData, error) {
	if strings.TrimSpace(authorizationHeader) == "" {
		return nil, apperror.NewUserError("AUTHORIZATION_HEADER_MISSING", "Authorization header is missing")
	}

	parts := strings.SplitN(authorizationHeader, " ", 2)
	if len(parts) != 2 || !strings.EqualFold(parts[0], "bearer") {
		return nil, apperror.NewUserError("AUTHORIZATION_HEADER_MALFORMATTED", "Authorization header is malformatted")
	}
	apiKey := strings.TrimSpace(parts[1])
	if len(apiKey) != a.cfg.IAM.APIKeyLength {
		return nil, apperror.NewUserError("AUTHORIZATION_HEADER_MALFORMATTED", "Authorization header is malformatted")
	}

	redisSession, err := a.sessionSvc.GetSessionByApiKey(ctx, apiKey)
	if err != nil {
		return nil, err
	}
	if redisSession == nil {
		return nil, apperror.NewUserError("API_KEY_EXPIRED", "Your session has expired. Login again.")
	}

	// Defense-in-depth expiration check using createdAt and configured validity duration.
	createdAt := time.UnixMilli(redisSession.CreatedAt)
	if time.Since(createdAt) > a.cfg.IAM.SessionValidityDuration {
		return nil, apperror.NewUserError("API_KEY_EXPIRED", "Your session has expired. Login again.")
	}

	user, err := a.userSvc.FindUserByIDOrFail(ctx, redisSession.UserID)
	if err != nil {
		return nil, err
	}

	return &model.AuthData{
		ApiKey:    apiKey,
		UserID:    redisSession.UserID,
		SessionID: redisSession.SessionID,
		User:      user,
	}, nil
}

